indexing-service/src/main/java/org/apache/druid/indexing/common/actions/RetrieveUsedSegmentsAction.java

```diff
@@ -113,11 +139,84 @@ public TypeReference<Collection<DataSegment>> getReturnTypeReference()
 
   @Override
   public Collection<DataSegment> perform(Task task, TaskActionToolbox toolbox)
+  {
+    if (!replace) {
```

Comment:
Cleaner to do:

```
if (isReplaceTask()) {
  return newMethodWhichDoesTheRightThingForReplaceTask();
} else {
  return retrieveUsedSegments(toolbox);
}

private boolean isReplaceTask() {
   return replace && task.getDatasource().equals(dataSource);
}
```

Reply:
The new method should also have a javadoc saying that it returns a consistent view of segments and why it is needed.

Reply:
I think the current structure makes it more readable as it is obvious that we are using the old logic whenever possible before trying to filter the locks.

```

if (datasourceToReplace != datasourceToRead) {
  return retrieveUsedSegments();
}

Set<ReplaceLock> replaceLocks = fetchReplaceLocksForTask();

if (replaceLocks.isEmpty()) {
  return retrieveUsedSegments()
} 

return retrieveUsedSegmentsCreatedBeforeReplaceVersions(replaceLocks);
```